lib/graphql/stitching/shaper.rb

```diff
@@ -0,0 +1,92 @@
+# typed: false
+# frozen_string_literal: true
+
+module GraphQL
+  module Stitching
+    class Shaper
+      def initialize(supergraph:, document:, raw:)
+        @supergraph = supergraph
+        @document = document
+        @result = raw
+        @errors = []
+      end
+
+      def perform!
+        if @result.key?("data") && ! @result["data"].empty?
+          munge_entry(@result["data"], @document.operation.selections, @supergraph.schema.query)
+          # hate doing a second pass, but cannot remove _STITCH_ fields until the fragements are processed
+          clean_entry(@result["data"])
+        end
+
+        if @errors.length > 0
+          @result = [] unless @result.key?("errors")
+          @result["errors"] += @errors
+        end
+
+        @result
+      end
+
+      private
+
+      def munge_entry(entry, selections, parent_type)
+        selections.each do |node|
+          case node
+          when GraphQL::Language::Nodes::Field
+            next if node.respond_to?(:name) && node&.name == "__typename"
```

Comment:
Now that we're doing this within `GraphQL::Language::Nodes::Field`, this condition can be simplified down to just...

```ruby
next if node.name == "__typename"
```

The class check assures that we have the expected name method.